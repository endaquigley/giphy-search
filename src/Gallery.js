import React, { useEffect } from "react";
import styled from "styled-components";
import { connect } from "react-redux";

import Modal from "./Modal.js";
import Thumbnail from "./Thumbnail.js";
import * as actions from "./actions";

const Wrapper = styled.div`
  margin: 15px;
  display: grid;
  grid-gap: 15px;
  --thumbnail-width: 85px;
  grid-template-columns: repeat(auto-fit, minmax(var(--thumbnail-width), 1fr));

  @media (min-width: 690px) {
    --thumbnail-width: 120px;
  }
`;

export const Gallery = React.memo(
  ({ data, page, query, selected, fetchImages, updateSelected }) => {
    useEffect(
      () => {
        fetchImages();
      },
      [query, page]
    );

    const renderSelectedModal = () => {
      if (selected) {
        return (
          <Modal
            selected={selected}
            handleClick={() => updateSelected(undefined)}
          />
        );
      }
    };

    const renderThumbnail = media => {
      if (media.images.preview_gif === undefined) {
        return;
      }

      return (
        <Thumbnail
          key={media.id}
          source={media.images.preview_gif.url}
          handleClick={() => updateSelected(media)}
        />
      );
    };

    return (
      <Wrapper>
        {renderSelectedModal()}
        {data.map(media => renderThumbnail(media))}
      </Wrapper>
    );
  }
);

const mapStateToProps = state => {
  return {
    data: state.data,
    page: state.page,
    query: state.query,
    selected: state.selected
  };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchImages: () => dispatch(actions.fetchImages()),
    updateSelected: selected => {
      return dispatch(actions.updateSelected(selected));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Gallery);
