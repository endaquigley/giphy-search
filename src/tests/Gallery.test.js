import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { render } from "react-testing-library";

import { Gallery } from "../Gallery";

const mockStore = configureStore([]);

describe("Gallery Component", () => {
  it("renders no image elements when data is empty", () => {
    const store = mockStore({
      data: []
    });

    const { container } = render(
      <Provider store={store}>
        <Gallery />
      </Provider>
    );

    const images = container.querySelectorAll("img");
    expect(images.length).toBe(0);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders an image element for each element in the data array", () => {
    const store = mockStore({
      data: [
        {
          id: 1,
          images: {
            original: { mp4: "https://enda.ie/filename-01.mp4" },
            preview_gif: { url: "https://enda.ie/filename-01.gif" }
          }
        },
        {
          id: 2,
          images: {
            original: { mp4: "https://enda.ie/filename-02.mp4" },
            preview_gif: { url: "https://enda.ie/filename-02.gif" }
          }
        }
      ]
    });

    const { container } = render(
      <Provider store={store}>
        <Gallery />
      </Provider>
    );

    const images = container.querySelectorAll("img");
    expect(images.length).toBe(2);

    expect(images[0].src).toBe("https://enda.ie/filename-01.gif");
    expect(images[1].src).toBe("https://enda.ie/filename-02.gif");

    expect(container.firstChild).toMatchSnapshot();
  });
});
