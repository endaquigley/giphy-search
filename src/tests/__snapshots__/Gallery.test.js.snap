// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery Component renders an image element for each element in the data array 1`] = `
<div>
  <div
    class="sc-bZQynM dNANyc"
  >
    <div
      class="sc-bxivhb hPtXaV"
    >
      <div
        class="sc-ifAKCX BwnNO"
      >
        <img
          class="sc-EHOje jjfgVa"
          src="https://enda.ie/filename-01.gif"
        />
      </div>
    </div>
    <div
      class="sc-bxivhb hPtXaV"
    >
      <div
        class="sc-ifAKCX BwnNO"
      >
        <img
          class="sc-EHOje jjfgVa"
          src="https://enda.ie/filename-02.gif"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Gallery Component renders no image elements when data is empty 1`] = `
<div>
  <div
    class="sc-bZQynM dNANyc"
  />
</div>
`;
